Step3: 카드덱 구현과 테스트

kjoonk opened the pull request:
## Deck 구현
- createAllCards(): suit 당 13장의 카드, 총 52(13 * 4)장의 카드를 생성해야 합니다. 반복문을 두 번 돌게 되는데, 이를 따로 분리했습니다. 조금은 덜 복잡해보일 것 같아서입니다. 이 구현을 Card 클래스 메소드로 옮기는 것도 잠시 고민했지만, Deck 쪽에 더 관련돼 있는 것 같아 그냥 두었습니다.
- cards: 덱이 가지고 있는 card들은 배열로 처리했습니다. 기본 컬렉션을 활용하는 게 더 간단해보일 것 같아서입니다.
- shuffle(), removeOne(), reset(): 모두 mutating이 되었습니다. shuffle의 경우, 실제 포커 게임에서 영구적으로 섞는 것이기 때문에 mutating으로 하였습니다.

kjoonk committed:
Card 클래스에 CustomStringConvertible 프로토콜 준수하도록 변경

kjoonk committed:
Deck 클래스 생성

- 카드 덱 관리용
- count, shuffle 등 구현해야 할 기능 선언만 함

kjoonk committed:
CardTest 추가

- 주요 메소드에 대해 테스트 추가: count, removeOne, reset

kjoonk committed:
Deck 클래스 구현

- count, shuffle, removeOne, reset 메소드 구현
- Card.Suit, Card.Rank에 CaseIterable 프로토콜 채택

godrm commented on CardGameApp/CardGameApp/Deck.swift:
```diff
@@ -0,0 +1,49 @@
+        }
+    }
+    
+    private func createCards(perSuit suit: Card.Suit) -> [Card] {
```
이런 경우는 `createCards(per suit: Card.Suit)`이라고 표현해도 충분히 읽기 쉽습니다. 참고하세요.

godrm commented on CardGameApp/CardGameApp/Deck.swift:
```diff
@@ -0,0 +1,49 @@
+    
+    /// 카드를 하나 반환하고 목록에서 삭제
+    mutating func removeOne() -> Card {
+        return cards.removeFirst()
```
cards.removeFirst()는 실패할 수 있지 않던가요? 배열이 없으면 어떻게 되나요?

kjoonk replied:
그렇네요. 오류보다는 옵셔널 처리가 나을 것 같습니다.

kjoonk replied:
옵셔널로 변경했습니다. 이 메소드를 호출하는 쪽에서 옵셔널 처리를 하는 것이 나아 보였어요.

godrm commented on CardGameApp/CardGameApp/Deck.swift:
```diff
@@ -0,0 +1,49 @@
+    private var cards = [Card]()
+    
+    init() {
+        self.cards = createAllCards()
```
Deck을 테스트하기 쉬운 구조로 만들기 위해서 
init() 함수를 `[Card]`매개변수를 받으면 어떨까요?
생성하는 부분을 분리하면 테스트가 조금 수월해지지 않을까요? 어떻게 생각하나요? 
그렇지 않으면 createAllCards()를 테스트할 방법이 없지는 않나요?

kjoonk replied:
음.. 고민해볼게요.
테스트할 방법이 없긴 했습니다.

kjoonk replied:
새 클래스를 만들어 Deck 생성 부분을 분리했습니다. 기존 Card나 Deck 클래스에는 어울리지 않아 보여서입니다.
이 여파로 reset 메소드도 변경을 해야 했습니다. 그래서 작동 방식을 바꿨어요. pop한 카드를 저장하고 있다가, reset할 때 cards에 복원하도록 했습니다.

godrm commented on CardGameApp/CardGameAppTests/CardTests.swift:
```diff
@@ -9,18 +9,29 @@
+        // reset() 테스트
+        deck.reset()
+        XCTAssertEqual(deck.count, numberOfAllCards)
+    }
```
기능별 함수가 여러 개인 경우는 각 함수별로 테스트 함수도 분리하세요.
그리고 각 기능이 성공하는 경우도 있지만, 실패하는 경우도 테스트 해야 합니다. 
이렇게 여러 단위 기능을 순서대로 호출해야 하는 경우는 단위 테스트가 아니라 통합 테스트라고 합니다. 별도 테스트 케이스에서는 통합 테스트를 만들어도 됩니다. 

테스트 함수를 분리해보세요.

kjoonk replied:
메소드 별로 분리했습니다.

godrm commented on CardGameApp/CardGameAppTests/CardTests.swift:
```diff
@@ -9,18 +9,29 @@
+        // count() 테스트
+        var deck = Deck()
+        let numberOfAllCards = 13 * 4
+        XCTAssertEqual(deck.count, numberOfAllCards)
```
위에 언급한 것처럼 초기 생성한 경우를 테스트하는 다른 방법도 고민해보세요.

godrm reviewed (commented):
다음 피드백을 확인해서 개선해보세요.

kjoonk committed:
Test 변경

- CardTest와 DeckTest로 분리
- DeckTest의 테스트 케이스를 메소드 별로 분리

kjoonk committed:
Deck 클래스 생성 및 작동 방식 변경

- DeckFactory 추가: Deck 생성 담당
- removeOne 작동 방식 변경: 뽑은 카드가 Optional이 되도록 변경
- poppedCards 속성 추가, reset 메소드 변경: pop한 카드를 배열로 담고 있으며, reset하면 cards 속성에 복원

kjoonk commented:
피드백 반영한 커밋 올렸습니다. 내용은 각 피드백에 답변으로 달았습니다.

godrm commented on CardGameApp/CardGameAppTests/DeckTests.swift:
```diff
@@ -0,0 +1,52 @@
+            }.count
+        }
+        
+        XCTAssertEqual(numberPerSuit, [13, 13, 13, 13])
```
여기 테스트들에서 자주 반복되는 13 이란 숫자도 하드코딩하기 보다는 코드로 표현해서 의미를 알 수 있도록 하는게 좋습니다. 참고하세요.

godrm commented on CardGameApp/CardGameAppTests/DeckTests.swift:
```diff
@@ -0,0 +1,52 @@
+
+class DeckTests: XCTestCase {
+    var deck: Deck!
+    let numberOfCards = 13 * 4
```
numberOfCards 가 13 * 4 만 될 수 있는걸까요? 
단어만 보면 전체 카드의 개수를 의미하는 게 아닌 것 같습니다.

godrm commented on CardGameApp/CardGameApp/Deck.swift:
```diff
@@ -38,12 +28,36 @@ struct Deck {
+    static func createCards(per suit: Card.Suit) -> [Card] {
+        return Card.Rank.allCases.map {
+            Card(suit: suit, rank: $0)
+        }
```
private 함수가 아니면 이 함수도 테스트 함수가 있어야 할 것 같네요.
